src/bip322.rs

```diff
@@ -0,0 +1,337 @@
+    /// created without a BIP322Signature. Such structure can only produce
+    /// to_spend (or empty to_sign) transaction, but cannot validate them.
+    signature: Option<Bip322Signature>,
+
+    /// script_pubkey to define the challenge script inside to_spend transaction
+    /// here we take in descriptors to derive the resulting script_pubkey
+    message_challenge: Descriptor<T>,
+}
+
+impl<T: MiniscriptKey + ToPublicKey> Bip322<T> {
+    /// Create a new BIP322 validator
+    pub fn new(msg: &[u8], sig: Option<Bip322Signature>, addr: Descriptor<T>) -> Self {
+        Bip322 {
+            message: msg.to_vec(),
+            signature: sig,
+            message_challenge: addr,
+        }
+    }
+
+    /// Insert Signature inside BIP322 structure
+    pub fn insert_sig(&mut self, sig: Bip322Signature) {
+        self.signature = Some(sig)
+    }
+
+    /// create the to_spend transaction
+    pub fn to_spend(&self) -> Transaction {
+        // create default input and output
+        let mut vin = TxIn::default();
+        let mut vout = TxOut::default();
+
+        // calculate the message tagged hash
+        let msg_hash = MessageHash::hash(&self.message[..]).into_inner();
+
+        // mutate the input with appropriate script_sig and sequence
+        vin.script_sig = Builder::new()
+            .push_int(0)
+            .push_slice(&msg_hash[..])
+            .into_script();
+        vin.sequence = 0;
+
+        // mutate the value and script_pubkey as appropriate
+        vout.value = 0;
+        vout.script_pubkey = self.message_challenge.script_pubkey();
+
+        // create and return final transaction
+        Transaction {
+            version: 0,
+            lock_time: 0,
+            input: vec![vin],
+            output: vec![vout],
+        }
+    }
+
+    /// Create to_sign transaction
+    /// This will create a transaction structure with empty signature and witness field
+    /// its up to the user of the library to fill the Tx with appropriate signature and witness  
+    pub fn to_sign(&self) -> Transaction {
+        // create the appropriate input
+        let outpoint = OutPoint::new(self.to_spend().txid(), 0);
+        let mut input = TxIn::default();
+        input.previous_output = outpoint;
+        input.sequence = 0;
+
+        // create the output
+        let output = TxOut {
+            value: 0,
+            script_pubkey: Builder::new()
+                .push_opcode(opcodes::all::OP_RETURN)
+                .into_script(),
+        };
+
+        // return resulting transaction
+        Transaction {
+            version: 0,
+            lock_time: 0,
+            input: vec![input],
+            output: vec![output],
+        }
+    }
+
+    /// Validate a BIP322 Signature against the message and challenge script
+    /// This will require a BIP322Signature inside the structure
+    pub fn validate(&self) -> Result<bool, BIP322Error> {
```

Comment:
We should reproduce the errors states from the BIP. This should return Result<(T, S), BIP322Error> where BIP322Error should either be `Inconclusive(..)` or `Invalid(..)`

Reply:
Is it possible to somehow derive these error types from the `IterpreterError`s?

Also if I understand it right, neither `to_spend` nor `to_sign` transactions are supposed to be mined. They are not validated against the chain. So what the purpose of `T` and `S` here and how the validator is supposed to validate `Time` and `Age` of an unmined transaction? Also, what does it mean to have a signature against a message only valid "after x time"? These are not transactions, but simple generic messages.